crates/krator/src/store.rs

```diff
@@ -0,0 +1,136 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use kube::api::DynamicObject;
+use kube::api::GroupVersionKind;
+use serde::de::DeserializeOwned;
+use tokio::sync::RwLock;
+
+use crate::object::ObjectKey;
+
+type ResourceMap = HashMap<GroupVersionKind, HashMap<ObjectKey, serde_json::Value>>;
+
+/// Defines Store type for caching Kubernetes objects locally.
+///
+/// * State is held in `Arc` so it is cheap to clone.
+/// * Collections are scoped by {group, version, kind, namespace, name}.
+/// * Objects are stored as [DynamicObject](kube::api::DynamicObject)s.
+///
+/// ```
+/// # use krator::Store;
+/// # use k8s_openapi::api::core::v1::Pod;
+/// #
+/// # #[tokio::main]
+/// # async fn main() -> anyhow::Result<()> {
+/// let store = Store::new();
+///
+/// let pod = Pod::default();
+///
+/// store.insert(Some("namespace"), "name", pod).await;
+///
+/// // Elsewhere
+/// let pod = store.get::<Pod>(Some("namespace"), "name").await?;
+///
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct Store {
+    objects: Arc<RwLock<ResourceMap>>,
+}
+
+impl Default for Store {
+    fn default() -> Self {
+        Store::new()
+    }
+}
+
+impl Store {
+    /// Initialize empty store.
+    pub fn new() -> Self {
+        Store {
+            objects: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Insert a new object.
+    pub async fn insert<R: 'static + k8s_openapi::Resource + Sync + Send + serde::Serialize>(
+        &self,
+        namespace: Option<&str>,
+        name: &str,
+        object: R,
+    ) {
+        let mut objects = self.objects.write().await;
+        let key = GroupVersionKind::gvk(R::GROUP, R::VERSION, R::KIND).unwrap();
+        let resource_objects = (*objects).entry(key).or_insert_with(HashMap::new);
+        let object_key = ObjectKey::new(namespace.map(|s| s.to_string()), name.to_string());
+        resource_objects.insert(object_key, serde_json::to_value(&object).unwrap());
+    }
+
+    /// Clear cache for specified object kind.
+    pub async fn reset(&self, gvk: &GroupVersionKind) {
+        let mut objects = self.objects.write().await;
+        let key = gvk.clone();
+        let resource_objects = (*objects).entry(key).or_insert_with(HashMap::new);
+        resource_objects.clear();
+    }
+
+    /// Delete a cached object.
+    pub async fn delete_any(
```

Comment:
Nit: Maybe just call this `delete`? `delete_any` made me think I'd find a `delete` as well

Reply:
My convention here is that `_any` indicates that they only have the `GroupVersionKind` and not the concrete type, a remnant from when I was using `Box<dyn Any>` to store things. This isn't really intended to be a public API so I will just make it private to avoid confusion.

Reply:
I will also rename to `_gvk`.